internal/k8s/controllers/frrk8s_config_controller.go

```diff
@@ -0,0 +1,139 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"reflect"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+
+	frrv1beta1 "github.com/metallb/frr-k8s/api/v1beta1"
+	frrk8s "go.universe.tf/metallb/internal/bgp/frrk8s"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+type frrk8sConfigEvent struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+}
+
+func (evt *frrk8sConfigEvent) DeepCopyObject() runtime.Object {
+	res := new(frrk8sConfigEvent)
+	res.Name = evt.Name
+	res.Namespace = evt.Namespace
+	return res
+}
+
+func NewFRRK8sConfigEvent() event.GenericEvent {
+	evt := frrk8sConfigEvent{}
+	evt.Name = "reload"
+	evt.Namespace = "frrk8sreload"
+	return event.GenericEvent{Object: &evt}
+}
+
+type FRRK8sReconciler struct {
+	client.Client
+	Logger               log.Logger
+	Scheme               *runtime.Scheme
+	NodeName             string
+	Namespace            string
+	Reload               chan event.GenericEvent
+	desiredConfiguration frrv1beta1.FRRConfiguration
+	sync.Mutex
+}
+
+func (r *FRRK8sReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	level.Info(r.Logger).Log("controller", "FRRK8sReconciler", "start reconcile", req.NamespacedName.String())
+	defer level.Info(r.Logger).Log("controller", "FRRK8sReconciler", "end reconcile", req.NamespacedName.String())
+	updates.Inc()
+
+	current := frrv1beta1.FRRConfiguration{}
+	err := r.Get(ctx, client.ObjectKey{Name: frrk8s.ConfigName(r.NodeName), Namespace: r.Namespace}, &current)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	r.Lock()
+	desired := r.desiredConfiguration.DeepCopy()
+	r.Unlock()
+
+	if reflect.DeepEqual(current.Spec, desired.Spec) {
+		return ctrl.Result{}, nil
+	}
+	desired.Spec.DeepCopyInto(&current.Spec)
+	current.Name = desired.Name
+	current.Namespace = desired.Namespace
```

Comment:
nit: this is a bit confusing/unnecessary , also I think it might make sense to let the session manager send a spec instead of the whole resource, and the ObjectMeta will be handled here. wdyt?

Reply:
it's not unnecessary, but I can provide a better way to make it not confusing.
It is required for the scenario where the object was not created yet, and so at this point current is an empty object (because the key was not set).
I don't like having the object half filled on one side and the other half (despite only name + namespace) on the other, I think it makes sense for the underlying implementation to generate the object as as whole and to decide for its name / namespace

Reply:
now it's better, what I didn't like was that both tried "taking control" over the metadata